src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/WorkspaceProjectStateChangeDetector.cs

```diff
@@ -153,6 +163,46 @@ internal virtual void Workspace_WorkspaceChanged(object sender, WorkspaceChangeE
             }
         }
 
+        // Internal for testing
+        internal bool IsPartialComponentClass(Document document)
+        {
+            if (!document.TryGetSyntaxRoot(out var root))
+            {
+                return false;
+            }
+
+            var classDeclarations = root.DescendantNodes().OfType<ClassDeclarationSyntax>();
+            if (!classDeclarations.Any())
+            {
+                return false;
+            }
+
+            if (!document.TryGetSemanticModel(out var semanticModel))
+            {
+                // This will occasionally return false resulting in us not refreshing TagHelpers for component partial classes. This means there are situations when a user's
+                // TagHelper definitions will not immediately update but we will eventually acheive omniscience.
+                return false;
+            }
+
+            var icomponentType = semanticModel.Compilation.GetTypeByMetadataName(ComponentsApi.IComponent.MetadataName);
```

Comment:
Hey @NTaylorMullen could this be null? I'm seeing this being null, and then ComponentDetectionConventions.IsComponent on line 197 blowing up with an ArgumentNullException.

Reply:
Oh damn, it can! That's definitely a bug. That being said I wouldn't expect it to pop up regularly because the default setup in 3.1 shouldn't hit that case unless something in your project was malformed. This makes me believe that the MonoDevelop compilation isn't properly setup

Reply:
But it's a regular C# project, there's no Razor anywhere in sight. Have you tried just debugging a console app with this? From what I can see we subscribe to any workspace change notification, and we don't filter by project type or otherwise...

Reply:
Hmm, that might be an issue with how VS4mac subscribes to project events then. In VS we don't load unless it's a Razor, for instance:

![image](https://user-images.githubusercontent.com/2008729/67594531-64dda080-f719-11e9-95a4-6ff3b43a17b5.png)